### ISSUE TITLE
'backup-dir' doesn't work when it contains a ~

### ISSUE THREAD

**Issue by joachim-n** [drush-ops/drush#1796]
I tried setting this in my drushrc.php:

```
$command_specific['pm-update'] = array('backup-dir' => "~/Library/Drush/drush-backups");
```

It appears to work in that no errors are shown when I try a code update:

```
Project features was updated successfully. Installed version is now 7.x-2.7.
Backups were saved into the directory                                [ok]
~/Library/Drush/drush-backups/drupal-joachim-foo/20151120170041/modules/features.
No database updates required                                         [success]
'all' cache was cleared.                                             [success]
Finished performing updates.                                         [ok]
$ open ~/Library/Drush/drush-backups/
$ open ~/Library/Drush/drush-backups/drupal-joachim-foo/20151120170041/modules/features
The file /Users/joachim/Library/Drush/drush-backups/drupal-joachim-foo/20151120170041/modules/features does not exist.
$ ls ~/Library/Drush/drush-backups/
$
```

If I change it to:

```
$command_specific['pm-update'] = array('backup-dir' => "/Users/joachim/Library/Drush/drush-backups");
```

then backup files are saved.

**Comment by weitzman** [comment]
We have drush_preflight_tilde_expansion() but it would take a long time to find all the paths one can specify in Drush and run this function on them. Any thoughts, @greg-1-anderson ?

**Comment by greg-1-anderson** [comment]
Yeah, the tilde expansion only runs on parameters typed on the CLI.  ~ is viewed as a shortcut whose purpose is to reduce the number of keys you have to type when entering commands.

For `drushrc.php` files, the intention was that you should not use this shortcut, but should instead say exactly what you mean in the script. For example:
```
$home = drush_server_home();
$command_specific['pm-update'] = array('backup-dir' => "$home/Library/Drush/drush-backups");
```
I wouldn't want to do tilde replacement too liberally, lest we accidentally substitute a ~ in some parameter that wasn't supposed to be a path. I don't think we'd really be doing anyone a favor if we started doing that in config files.

**Comment by webflo** [comment]
FYI, https://github.com/webmozart/path-util is part of Puli and has a few helper for path handling.

**Comment by joachim-n** [comment]
> I wouldn't want to do tilde replacement too liberally, lest we accidentally substitute a ~ in some parameter that wasn't supposed to be a path.

I was wondering whether we could define in hook_drush_command() whether parameters and options need this, but that would mean changing the API to have an array of data for each one rather than just the description string.

**Comment by greg-1-anderson** [comment]
Options can already be an array. If an option is a string, it is converted to `array('description' => STRING)`.  We could potentially add an attribute to indicate the format of an option, and use that to flag paths.

**Comment by greg-1-anderson** [comment]
path-util++
